src/XML/src/FMI3/fmi3_xml_variable.c

```diff
@@ -968,19 +965,85 @@ static void* fmi3_xml_get_variable_start_array(fmi3_xml_variable_t* v) {
+        // with 2 each time, we therefore would just ignore the invalid G. Note that we would get a
+        // failure if the input was instead "GF".
+        //
+        // So while it's enough to check the second character, we may as well check all for consistency.
+        for (size_t j = 0; j <= 1; j++) {
+            char ch = pos[j];
+            if (!((ch >= '0' && ch <= '9') || (ch >= 'a' && ch <= 'f') || (ch >= 'A' && ch <= 'F'))) {
+                fmi3_xml_parse_error(context, "String is not hexadecimal: %s", hexstr);
+                return -1;
+            }
+        }
+
+        if (!sscanf(pos, "%2hhx", &bytearr[i])) {
+            fmi3_xml_parse_error(context, "Failure when scanning hexadecimal string: %s", hexstr);
+            return -1;
+        }
+        pos += 2;
+    }
+    return 0;
+}
+
 void fmi3_xml_variable_free_internals(jm_callbacks* callbacks, fmi3_xml_variable_t* var) {
+    /* We need special handling of Strings and Binaries because
+    *  of how start values are handled differently for those types.
+    */
     if (fmi3_xml_variable_is_array(var)) {
         if (var->type->baseType == fmi3_base_type_str) {
             fmi3_xml_string_variable_start_t* start = (fmi3_xml_string_variable_start_t*)(var->type);
-            if (&start->arrayStartValues) {
-                jm_vector_foreach(jm_voidp)(&start->arrayStartValues, callbacks->free);
-                jm_vector_free_data(jm_voidp)(&start->arrayStartValues);
+            if (&start->stringStartValues) {
+                jm_vector_foreach(jm_voidp)(&start->stringStartValues, callbacks->free);
+                jm_vector_free_data(jm_voidp)(&start->stringStartValues);
+            }
+        } else if (var->type->baseType == fmi3_base_type_binary) {
+            fmi3_xml_binary_variable_start_t* start = (fmi3_xml_binary_variable_start_t*)(var->type);
+            if (&start->binaryStartValues) {
+                jm_vector_foreach(jm_voidp)(&start->binaryStartValues, callbacks->free);
+                jm_vector_free_data(jm_voidp)(&start->binaryStartValues);
+            }
+            if (&start->binaryStartValuesSize) {
+                jm_vector_free_data(size_t)(&start->binaryStartValuesSize);
             }
         } else {
             callbacks->free(fmi3_xml_get_variable_start_array(var));
         }
         callbacks->free(var->dimensionsArray);
         var->dimensionsArray = NULL;
+    } else {
+        /* Scalar String variables can have a start value that we need to free.*/
+        if (var->type->baseType == fmi3_base_type_str) {
+            fmi3_xml_string_variable_start_t* start = (fmi3_xml_string_variable_start_t*)(var->type);
+            if (&start->stringStartValues) {
+                jm_vector_foreach(jm_voidp)(&start->stringStartValues, callbacks->free);
+                //jm_vector_free_data(jm_voidp)(&start->stringStartValues);
+            }
+        }
```

Comment:
We treat it differently compared to binary because Binary variables that are not arrays are using the old implementation (because of the reasons we discussed when we realized that errors during parsing are treated very differently from case to case).

jm_vector_free_data here causes `Invalid read of size 8` and since memcheck does not complain **without** `jm_vector_free_data` I was initially thinking the current implementation is correct.